Expand log level extension

ChrisSimmons opened the pull request:
This PR expands on #629.  Changes:

* Adds a new extension method that also accepts an `EventId` and an `Exception`
* Update [the previous method](https://github.com/ChrisSimmons/Logging/blob/a4b901159255abbdd47cf979171838799f971f66/src/Microsoft.Extensions.Logging.Abstractions/LoggerExtensions.cs#L350) to use this new method
* Update all the level-specific methods (e.g. [`LogDebug()`](https://github.com/ChrisSimmons/Logging/blob/a4b901159255abbdd47cf979171838799f971f66/src/Microsoft.Extensions.Logging.Abstractions/LoggerExtensions.cs#L42) and [`LogWarning()`](https://github.com/ChrisSimmons/Logging/blob/a4b901159255abbdd47cf979171838799f971f66/src/Microsoft.Extensions.Logging.Abstractions/LoggerExtensions.cs#L191)) to use this new method
* Use the new method to [handle the null `logger` argument exception](https://github.com/ChrisSimmons/Logging/blob/a4b901159255abbdd47cf979171838799f971f66/src/Microsoft.Extensions.Logging.Abstractions/LoggerExtensions.cs#L366)
    * If this is not desirable, e.g. for stacktrace reasons, I can revert and resubmit the PR

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/aspnet/Logging?pullRequest=790) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/aspnet/Logging?pullRequest=790) before we can accept your contribution.<br/><br/>:x: ChrisSimmons [sign now](https://cla.dotnetfoundation.org/aspnet/Logging?pullRequest=790)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/aspnet/Logging?pullRequest=790) it.</sub>

Eilon commented:
@muratg - do we want to have someone take a look at this?

muratg commented:
@ChrisSimmons thanks for the PR!

@pakrym could you review when you have some time?

pakrym commented on test/Microsoft.Extensions.Logging.Test/LoggerExtensionsTest.cs:
```diff
@@ -38,10 +38,15 @@ public void MessageOnly_LogsCorrectValues()
             logger.LogError(_state);
             logger.LogCritical(_state);
             logger.LogDebug(_state);
+            logger.Log(LogLevel.Trace, _state);
```
Add a separate test method that calls new overloads (with exception and non zero event id)

ChrisSimmons replied:
Sorry for the delay.  I'll get to this now.

ChrisSimmons replied:
I've added separate tests for the new overloads, one for each of the 10 "flavors" of the equivalent `logger.LogLEVELNAME()` extensions.  Each uses a `LogLevel_` prefix to the test method name ([Example](https://github.com/aspnet/Logging/pull/790/files#diff-099caadefe774232f5d421983c3ee853R627)).  Let me know if that naming works.

(FYI, I rebased and squashed commits, @pakrym.)

pakrym commented:
Please add tests for new overloads, otherwise LGTM.

ChrisSimmons commented:
I rebased and squashed commits, which blew away your previous comments, @pakrym:

> Add a separate test method that calls new overloads (with exception and non zero event id)

I've added separate tests for the new overloads, one for each of the 10 "flavors" of the equivalent `logger.LogLEVELNAME()` extensions.  Each uses a `LogLevel_` prefix to the test method name ([Example](https://github.com/aspnet/Logging/pull/790/files#diff-099caadefe774232f5d421983c3ee853R627)).  Let me know if that naming works.

ChrisSimmons commented:
Also, is there any value to using a `[Theory]` to simplify these tests?  See [this](https://github.com/ChrisSimmons/Logging/commit/a113909bda2c685938e0c80a4469add8d364afcb?diff=split), for example.

pakrym commented:
Totally, bring the Theories in.

pakrym commented on test/Microsoft.Extensions.Logging.Test/LoggerExtensionsTest.cs:
```diff
@@ -628,6 +623,267 @@ public void LogValuesEventIdAndError_LogsCorrectValues()
+            logger.Log(logLevel, eventId, _exception, testLogValues.ToString());
+
+            // Assert
+            Assert.True(sink.Writes.TryTake(out var write));
```
Assert.Single(sink.Writes);
